aioshelly/block_device.py

```diff
@@ -89,54 +95,72 @@ def ip_address(self) -> str:
         """Device ip address."""
         return self.options.ip_address
 
-    async def initialize(self) -> None:
+    async def initialize(self, async_init: bool = False) -> None:
         """Device initialization."""
+        if self._initializing:
+            raise RuntimeError("Already initializing")
+
         self._initializing = True
         self.initialized = False
+        ip = self.options.ip_address
         try:
             await self.update_shelly()
 
-            if self.options.auth or not self.requires_auth:
+            if self.requires_auth and not self.options.auth:
+                raise InvalidAuthError("auth missing and required")
+
+            async with async_timeout.timeout(DEVICE_IO_TIMEOUT):
                 await self.update_settings()
                 await self.update_status()
 
-            event_d: asyncio.Event = await self.coap_request("d")
+                event_d: asyncio.Event = await self._coap_request("d")
+                # We need to wait for D to come in before we request S
+                # Or else we might miss the answer to D
+                await event_d.wait()
 
-            # We need to wait for D to come in before we request S
-            # Or else we might miss the answer to D
-            await event_d.wait()
-
-            if self._request_s:
-                event_s = await self.coap_request("s")
-                await event_s.wait()
+                if not async_init:
+                    event_s = await self._coap_request("s")
+                    await event_s.wait()
 
             self.initialized = True
+        except ClientResponseError as err:
+            if err.status == HTTPStatus.UNAUTHORIZED:
+                self._last_error = InvalidAuthError(err)
+                _LOGGER.debug("host %s: error: %r", ip, self._last_error)
+                # Auth error during async init, used by sleeping devices
+                # Will raise 'invalidAuthError; on next property read
+                if not async_init:
+                    self.shutdown()
+                    raise InvalidAuthError(err) from err
+                self.initialized = True
+            else:
+                raise DeviceConnectionError from err
```

Comment:
Why do we not store the error here as last error?

Reply:
Yes we should store the last error and looking at the code again we should not raise if it is async init since there is no caller to catch it